### ISSUE TITLE
Add GPG signatures or file hashes to released files

### ISSUE THREAD

**Issue by marinbernard-pep06** [bdd/runitor#25]
Hi,

Currently, each _runitor_ release comes with source code and binaries for various platforms (thanks for that!). Unfortunately, people downloading those files have no way to check their integrity/authenticity.

Would you mind including GPG signatures (or at least files hashes) for every file in the next releases ?

Thank you

**Comment by bdd** [comment]
## Re: detached signature files
I will not use PGP. I may consider using [Signify](https://man.openbsd.org/signify.1) or [Minisign]() though.

Regardless of implementation, this will be a manual process. Right now release binaries are compiled and attached to releases with [a GitHub Action](https://github.com/bdd/runitor/blob/main/.github/workflows/release.yml). Thanks to Go's reproducible builds, I'll do a compilation on a trusted host, and generate the signature file on a trusted host.
I think if I can just yank the `buildid` out of linked binaries and we'll be good to go for fully reproducible builds.

## Re: unsigned checksum files
I can provide them but help me understand for what purpose you intend to rely on them.
It's rather important to not crate a false sense of security/assurance.

### They cannot verify authenticity
As I explained above, releases are compiled and uploaded by GH Actions. If this delivery pipeline or even worse my GitHub account is compromised, these files will be moot.

### Don't need them to catch network issues
Granted binary downloads are only available over TLS, the integrity is assured end-to-end. The only cases a [CHF](https://en.wikipedia.org/wiki/Cryptographic_hash_function) based checksum file would help is a data corruption issue that happens _before_ [TLS recordization](https://www.oreilly.com/library/view/the-ims-ip/9780470019061/9780470019061_tls_record_protocol.html#:~:text=The%20TLS%20Record%20Protocol%20layers,Message%20Authentication%20Checksum%20(MAC).) on the sending end, **or** _after_ TLS stack processing on the receiving end.   So, realistically they _may_ help catching fringe corruption issues on the network stacks of either end, or a corruption at rest (in situ) at GitHub's S3 bucket, or some filesystem issue on the receiving end.

All in all, unsigned checksum files doesn't provide anything about authenticity, and does very little for integrity at the age of ubiquitous TLS.

**Comment by bdd** [comment]
Added SHA256 generation to the distribution build script in https://github.com/bdd/runitor/commit/9b51f81bad86e626dcb0d849afe86a0815dece29.

...and retroactively [added to v0.8.0 artifacts](https://github.com/bdd/runitor/releases/download/v0.8.0/SHA256)

I will  commit instructions on how to verify SHA256.sig files with Signify, in another commit, hopefully this week.

**Comment by marinbernard-pep06** [comment]
> All in all, unsigned checksum files doesn't provide anything about authenticity, and does very little for integrity at the age of ubiquitous TLS.

Absolutely. I only mentioned them as a poor man fallback if GPG/signify would not be considered. Signify is great.

Thank you for taking the time for this.

**Comment by bdd** [comment]
[SHA256.sig for v0.8.0](https://github.com/bdd/runitor/releases/download/v0.8.0/SHA256.sig) and [SHA256 for 0.9.0-beta.1](https://github.com/bdd/runitor/releases/download/v0.9.0-beta.1/SHA256.sig) are added.

I couldn't find time to write verification instructions tonight, but essentially my plan is to distribute the public keys from a canonical location hosted on the same domain of the Go module vanity URL of runitor. I plan to rotate the signing key every major version or increment the major version if there was a key compromise. So far I've generated keys for v0 (https://bdd.fi/sig/runitor-v0.pub) and v1 (https://bdd.fi/sig/runitor-v1.pub) major versions. If you (or anyone reading this) have suggestions for an improved process, I'm all ears.

If you're curious about what my workflow for signing looks like, [helper scripts are in the 'autograph' branch](https://github.com/bdd/runitor/tree/autograph/scripts) now, pending to be merged to master as soon as I write another markdown file, describing how to verify the authenticity of release binaries; both en masse, or individually.

**Comment by ofer-dev** [comment]
You may want to look at [sigstore](https://www.sigstore.dev/) for signing artifacts.
